### PR TITLE
Python: Introduced versioned api directory exporting types.

### PR DESCRIPTION
**Pull request by landrito**
Updates #1320

### CODE REVIEW

File: src/main/java/com/google/api/codegen/transformer/SurfaceNamer.java [hunk 1]
```diff
@@ -159,6 +159,20 @@ public String getNamespace(Interface apiInterface) {
     return qualifiedName(namePath.withoutHead());
   }
 
+  /** The namespace of the package. */
```
Comment (geigerj) [inline src/main/java/com/google/api/codegen/transformer/SurfaceNamer.java]: nit: "the _unversioned_ namespace..."

File: src/main/java/com/google/api/codegen/transformer/SurfaceNamer.java [hunk 2]
```diff
@@ -475,6 +489,15 @@ public String getByteLengthFunctionName(TypeRef typeRef) {
     return getNotImplementedString("SurfaceNamer.getByteLengthFunctionName");
   }
 
+  /** The name of the class that ties in the helper functions to a client partial veneers. */
+  public String getHelpersClassName(String apiShortName) {
+    return publicClassName(Name.anyCamel(apiShortName).join("helpers"));
+  }
+
+  /** The name of a stub type of an interface */
+  public String getStubType(Interface apiInterface) {
```
Comment (geigerj) [inline src/main/java/com/google/api/codegen/transformer/SurfaceNamer.java]: How is this different from the value we pass to `main.snip` to instantiate the gRPC stub?

File: src/main/java/com/google/api/codegen/transformer/py/PythonImportSectionTransformer.java [hunk 3]
```diff
@@ -151,6 +183,101 @@ private ImportFileView generateApiImport(GapicMethodContext context) {
...
+      // For python, adding a single message from the proto file to the type table will populate
+      // the type table with the correct imports.
+      ImmutableList<MessageType> messages = protoFile.getMessages();
+      if (!messages.isEmpty()) {
+        typeTable.getAndSaveNicknameFor(TypeRef.of(messages.get(0)));
+      }
+    }
+  }
+
+  private Set<ProtoFile> getProtofileDependencies(Model model) {
```
Comment (geigerj) [inline src/main/java/com/google/api/codegen/transformer/py/PythonImportSectionTransformer.java]: optional: it is probably significantly less efficient, but the simpler way to do this would be to subclass the api-compiler [Visitor](https://github.com/googleapis/api-compiler/blob/cc52101f81eae31ebdac647f9664f5b32d2228a1/src/main/java/com/google/api/tools/framework/model/Visitor.java). Here's [a more complicated example](https://github.com/googleapis/api-compiler/blob/4621e9bb63a3903b6b2c4639f6560c9284585a0f/src/main/java/com/google/api/tools/framework/processors/resolver/SymbolTableBuilder.java#L45) of doing this; in your case, the subclass would be trivial (it just needs to add ProtoFiles to a Set when it visits them).

File: src/main/java/com/google/api/codegen/transformer/py/PythonImportSectionTransformer.java [hunk 4]
```diff
@@ -151,6 +183,101 @@ private ImportFileView generateApiImport(GapicMethodContext context) {
...
+      SurfaceNamer namer) {
+
+    Iterable<Interface> apiInterfaces = new InterfaceView().getElementIterable(model);
+    ModelTypeTable typeTable = emptyTypeTable(productConfig);
+    for (Interface apiInterface : apiInterfaces) {
+      namer.getAndSaveNicknameForGrpcClientTypeName(typeTable, apiInterface);
+    }
+
+    Set<ImportFileView> imports = new TreeSet<>(importFileViewComparator());
+    // TODO: (landrito) Make this only generate the enums module if there are enums to export.
```
Comment (geigerj) [inline src/main/java/com/google/api/codegen/transformer/py/PythonImportSectionTransformer.java]: [Easy check](https://github.com/googleapis/toolkit/blob/f0762cd69fa4286fca3cfe7445b369c887ba5061/src/main/java/com/google/api/codegen/py/PythonImportHandler.java#L83). It seems like this was cause an `ImportError` if you actually try to run a GAPIC with no enums, so it seems like it would be best just to address this now.

File: src/main/java/com/google/api/codegen/transformer/py/PythonSurfaceNamer.java [hunk 5]
```diff
@@ -146,8 +146,8 @@ public String getParseFunctionName(String var, SingleResourceNameConfig resource
 
   @Override
   public String getGrpcClientTypeName(Interface apiInterface) {
-    String fullName = getModelTypeFormatter().getFullNameFor(apiInterface) + "Stub";
-    return getTypeNameConverter().getTypeName(fullName).getNickname();
+    String fullName = getModelTypeFormatter().getFullNameFor(apiInterface) + "Client";
```
Comment (geigerj) [inline src/main/java/com/google/api/codegen/transformer/py/PythonSurfaceNamer.java]: Confused about terminology: `grpcClientTypeName` vs `stubType`?

File: src/main/java/com/google/api/codegen/transformer/SurfaceNamer.java (hunk 2 above)
Comment (landrito) [inline src/main/java/com/google/api/codegen/transformer/SurfaceNamer.java]: It's no different. The full stub type was stored in the property that held grpcClientTypeName. Which didn't seem correct. This was to fix that so that I could use the namer.getGrpcClientTypeName for the grpc client type name.

File: src/main/java/com/google/api/codegen/transformer/py/PythonImportSectionTransformer.java (hunk 3 above)
Comment (landrito) [inline src/main/java/com/google/api/codegen/transformer/py/PythonImportSectionTransformer.java]: OH LOL. I found an even simple solution. `Model` has the method `getFiles`.

File: src/main/java/com/google/api/codegen/transformer/py/PythonSurfaceNamer.java (hunk 5 above)
Comment (landrito) [inline src/main/java/com/google/api/codegen/transformer/py/PythonSurfaceNamer.java]: This change is actually to make it such that the terminology is intuitive. The grpcClientTypeName is the type of the grpcClient. The stubType is the type of the grpcClientStub.

File: src/main/java/com/google/api/codegen/transformer/SurfaceNamer.java (hunk 1 above)
Comment (landrito) [inline src/main/java/com/google/api/codegen/transformer/SurfaceNamer.java]: Done.

File: src/main/java/com/google/api/codegen/transformer/py/PythonImportSectionTransformer.java (hunk 4 above)
Comment (landrito) [inline src/main/java/com/google/api/codegen/transformer/py/PythonImportSectionTransformer.java]: Fixed. Had to include that data elsewhere so moved the check into the transformer.

File: src/test/java/com/google/api/codegen/testdata/python_doc_types_library.baseline [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+               descriptor_pb2
+               duration_pb2
+               empty_pb2
+               protobuf_field_mask_pb2
+               struct_pb2
+               timestamp_pb2
+               wrappers_pb2
+               status_pb2):
+    for name, message in get_messages(module).items():
+        message.__module__ = 'google.cloud.proto.example.library_v1'
```
Comment (lukesneeringer) [inline src/test/java/com/google/api/codegen/testdata/python_doc_types_library.baseline]: This line is _probably_ wrong? I would expect it to be `google.cloud.library_v1`.

File: src/test/java/com/google/api/codegen/testdata/python_doc_types_no_path_templates.baseline [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+from google.protobuf import empty_pb2
+
+
+names = []
+for module in (http_pb2
+               no_path_templates_pb2
+               descriptor_pb2
+               empty_pb2):
+    for name, message in get_messages(module).items():
+        message.__module__ = 'google.cloud.proto.example_v1'
```
Comment (lukesneeringer) [inline src/test/java/com/google/api/codegen/testdata/python_doc_types_no_path_templates.baseline]: I would expect this to be `message.__module__ = 'google.cloud.example_v1'`. But I can not tell if it is wrong since it is hard to determine that from the bogus examples.

File: src/test/java/com/google/api/codegen/testdata/python_doc_types_library.baseline [hunk 8]
```diff
@@ -0,0 +1,65 @@
+============== file: google/cloud/proto/example/library_v1/types.py ==============
```
Comment (lukesneeringer) [inline src/test/java/com/google/api/codegen/testdata/python_doc_types_library.baseline]: I would have expected this file location to be `google/cloud/library_v1/types.py`.

File: src/test/java/com/google/api/codegen/testdata/python_doc_types_no_path_templates.baseline [hunk 9]
```diff
@@ -0,0 +1,39 @@
+============== file: google/cloud/proto/example_v1/types.py ==============
```
Comment (lukesneeringer) [inline src/test/java/com/google/api/codegen/testdata/python_doc_types_no_path_templates.baseline]: I would have expected this file to be `google/cloud/example_v1/types.py`.

File: src/test/java/com/google/api/codegen/testdata/python_doc_versioned_directory__init___library.baseline [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+
+from __future__ import absoulte_import
+
+from google.cloud.gapic.example.library.helpers import LibraryHelpers
+from google.cloud.gapic.example.library.v1 import enums
+from google.cloud.proto.example.library.v1 import library_pb2
+from google.cloud.proto.example.library_v1 import types
+
+
+class LibraryServiceClient(LibraryHelpers, library_pb2.LibraryServiceClient):
```
Comment (lukesneeringer) [inline src/test/java/com/google/api/codegen/testdata/python_doc_versioned_directory__init___library.baseline]: `LibraryHelpers` is manually written and probably should not be referenced by autogen.

File: src/test/java/com/google/api/codegen/testdata/python_doc_versioned_directory__init___no_path_templates.baseline [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+
+
+from __future__ import absoulte_import
+
+from example.helpers import LibraryHelpers
+from google.cloud.proto.example.v1 import no_path_templates_pb2
+from google.cloud.proto.example_v1 import types
+
+
+class NoTemplatesAPIServiceClient(LibraryHelpers, no_path_templates_pb2.NoTemplatesAPIServiceClient):
```
Comment (lukesneeringer) [inline src/test/java/com/google/api/codegen/testdata/python_doc_versioned_directory__init___no_path_templates.baseline]: `LibraryHelpers` is manually written and likely should not be referenced by autogen.

File: src/test/java/com/google/api/codegen/testdata/python_types_library.baseline [hunk 12]
```diff
@@ -0,0 +1,65 @@
+============== file: google/cloud/proto/example/library_v1/types.py ==============
```
Comment (lukesneeringer) [inline src/test/java/com/google/api/codegen/testdata/python_types_library.baseline]: Same issue with before re: the file path (and __module__ below)

File: src/test/java/com/google/api/codegen/testdata/python_doc_types_library.baseline (hunk 8 above)
Comment (landrito) [inline src/test/java/com/google/api/codegen/testdata/python_doc_types_library.baseline]: I see. Hmm I'll need to think more about how to construct that path.

Comment (landrito) [inline src/test/java/com/google/api/codegen/testdata/python_doc_types_library.baseline]: @geigerj maybe you have an idea.

I implemented it using the gapic package name but that produced:
google/cloud/**gapic**/example/library_v1/

Now I'm thinking to construct it from the package config shortname name and version but I am trying to figure out how to get the base path.
as in
{base/path}/{shortname}_{version}

Thoughts?

Comment (geigerj) [inline src/test/java/com/google/api/codegen/testdata/python_doc_types_library.baseline]: What is `base/path`? Is it ever something other than `google/cloud`?

Comment (landrito) [inline src/test/java/com/google/api/codegen/testdata/python_doc_types_library.baseline]: It will if we ever move towards non cloud apis.

Comment (geigerj) [inline src/test/java/com/google/api/codegen/testdata/python_doc_types_library.baseline]: It might make sense to use the [`organization_name`](https://github.com/googleapis/googleapis/blob/master/gapic/api/artman_datastore.yaml#L6) configuration, then. (Just replace the hyphen with the file separator.) You probably will need to modify the [package metadata generation task](https://github.com/googleapis/artman/blob/master/artman/tasks/package_metadata_tasks.py#L26) in artman to pass this through to toolkit.

Comment (landrito) [inline src/test/java/com/google/api/codegen/testdata/python_doc_types_library.baseline]: I think using `organization_name` in that manner seems a bit magical.

I think a more ideal solution would be to have a property in the gapic yaml that holds the gapic package name and a property that holds the top-level package name. I think we should lean towards making these package names explicit rather than parsing the name out of an existing property.

ex:
```yaml
language_settings:
  python:
  package_name: google.cloud.ex_v1
  gapic_package_name: google.cloud.ex_v1.gapic
  proto_package_name: google.cloud.ex_v1.proto
```
Thoughts?

Comment (geigerj) [inline src/test/java/com/google/api/codegen/testdata/python_doc_types_library.baseline]: I might be missing the full context of this re-organization. Are the `google.cloud.{gapic,proto}.ex.v1` packages going away and being replaced by these `ex_v1.{gapic,proto}` packages?

Comment (landrito) [inline src/test/java/com/google/api/codegen/testdata/python_doc_types_library.baseline]: Ah yes but I think that it is orthogonal to this decision.

Using the current pattern the yaml would look like.

```
language_settings:
  python:
    package_name: google.cloud.example_v1
    gapic_package_name: google.cloud.gapic.example.v1
    proto_package_name: google.cloud.proto.example.v1

Comment (geigerj) [inline src/test/java/com/google/api/codegen/testdata/python_doc_types_library.baseline]: Ack. I'm not sure there is utility in having `gapic_package_name` and `proto_package_name` if they are always derivable from `package_name`. But just changing the semantics of `package_name` SGTM.

Comment (landrito) [inline src/test/java/com/google/api/codegen/testdata/python_doc_types_library.baseline]: Ack. 

@lukesneeringer how does that sound to you? Simply changing the semantics of the package name to be of the form `google.cloud.example_v1` and having the gapic and proto paths just added to that package name.

Comment (lukesneeringer) [inline src/test/java/com/google/api/codegen/testdata/python_doc_types_library.baseline]: > I might be missing the full context of this re-organization. Are the `google.cloud.{gapic,proto}.ex.v1` packages going away and being replaced by these `ex_v1.{gapic,proto}` packages?

They do not have to be necessarily, but it matters that the ultimate names we expose to users are `google.cloud.example_v1` and only things underneath that.

The current implementation in Vision, Speech, NL, etc. has modules that import from the existing namespaces, but changing the namespaces would actually be easier for documentation and such.

Comment (lukesneeringer) [inline src/test/java/com/google/api/codegen/testdata/python_doc_types_library.baseline]: > @lukesneeringer how does that sound to you? Simply changing the semantics of the package name to be of the form google.cloud.example_v1 and having the gapic and proto paths just added to that package name.
  
That is fine by me. It might even be ideal by me. :-)

Comment (lukesneeringer) [inline src/test/java/com/google/api/codegen/testdata/python_doc_types_library.baseline]: Addendum: I am now of the understanding that the new package names will be:

This means the final paths will be:

```yaml
package_name: google.cloud.example_v1
gapic_package_name: google.cloud.example_v1.gapic
proto_package_name: google.cloud.example_v1.proto
```

But only the first will be explicitly configured. Is that accurate?

Comment (lukesneeringer) [inline src/test/java/com/google/api/codegen/testdata/python_doc_types_library.baseline]: To be clear, the back-incompat change is fine. I am willing to go update the world to get this change. :-)

Comment (landrito) [inline src/test/java/com/google/api/codegen/testdata/python_doc_types_library.baseline]: > But only the first will be explicitly configured. Is that accurate?
Correct.

Comment (landrito) [inline src/test/java/com/google/api/codegen/testdata/python_doc_types_library.baseline]: Took a look at what it's gonna take to do the package name switch. It's a better fit to do that work in a separate PR.
